File: common_core/src/main/java/com/xueh/comm_core/net/ServiceGenerator.java
package com.xueh.comm_core.net;


import com.xueh.comm_core.helper.DevelopHelper;

import java.util.HashMap;
import java.util.Map;

import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

public class ServiceGenerator {

    private static Map<String, Retrofit> mServiceMap = new HashMap();

    private static final String DOMAIN_BASE = DevelopHelper.getBaseUrl();

    private ServiceGenerator() {
    }

    public static <T> T getService(Class<T> serviceClass) {
        return getCustomService(DOMAIN_BASE, serviceClass);
    }

    /**
     * @param domain Retrofit的BaseUrl
     */
    public static <T> T getCustomService(String domain, Class<T> serviceClass) {
        synchronized (ServiceGenerator.class) {
            Retrofit retrofit = mServiceMap.get(domain);
            if (retrofit == null) {
                retrofit = getRetrofit(domain);
                //只缓存最常用的
                if (DOMAIN_BASE.equals(domain)) {
                    mServiceMap.put(domain, retrofit);
                }
            }
            return createServiceFrom(retrofit, serviceClass);
        }
    }


    private static <T> T createServiceFrom(Retrofit retrofit, Class<T> serviceClass) {
        return retrofit.create(serviceClass);
    }

    private static Retrofit getRetrofit(String base_url) {
        return new Retrofit.Builder()
                .baseUrl(base_url)
                .client(HttpUtils.getOkHttp())
             //   .addConverterFactory(ScalarsConverterFactory.create())
                .addConverterFactory(GsonConverterFactory.create())         //返回内容的转换器
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())  //请求Call的转换器
                .build();
    }

    public static void reset() {
        mServiceMap.clear();
    }
}

File: common_core/src/main/java/com/xueh/comm_core/net/ServiceGenerator.kt
package com.xueh.comm_core.net

import io.github.iamyours.wandroid.net.LiveDataCallAdapterFactory
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.util.*

object ServiceGenerator {
    private val mServiceMap: MutableMap<String, Retrofit> = HashMap()
    private var DOMAIN_BASE=""
    fun setBaeUrl(httpBaseUrl: IHttpBaseUrl) {
        DOMAIN_BASE = httpBaseUrl.getBaseUrl()
    }

    fun <T> getService(serviceClass: Class<T>): T {
        return getCustomService(DOMAIN_BASE, serviceClass)
    }

    /**
     * @param domain Retrofit的BaseUrl
     */
    fun <T> getCustomService(domain: String, serviceClass: Class<T>): T {
        synchronized(ServiceGenerator::class.java) {
            var retrofit = mServiceMap[domain]
            if (retrofit == null) {
                retrofit = getRetrofit(domain)
                //只缓存最常用的
                if (DOMAIN_BASE == domain) {
                    mServiceMap[domain] = retrofit
                }
            }
            return createServiceFrom(retrofit, serviceClass)
        }
    }

    private fun <T> createServiceFrom(retrofit: Retrofit, serviceClass: Class<T>): T {
        return retrofit.create(serviceClass)
    }

    private fun getRetrofit(base_url: String): Retrofit {
        return Retrofit.Builder()
            .baseUrl(base_url)
            .client(HttpUtils.getOkHttp()) //   .addConverterFactory(ScalarsConverterFactory.create())
            .addConverterFactory(GsonConverterFactory.create()) //返回内容的转换器
            .addCallAdapterFactory(LiveDataCallAdapterFactory())
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create()) //请求Call的转换器
            .build()
    }

    fun reset() {
        mServiceMap.clear()
    }
}